### PR TITLE
Add .hpp header and deprecate .h

### PR DESCRIPTION
**Pull request by sloretz**
This PR is two commits and should be rebase and merged to capture the moved file in history. The first commit renames `retriever.h` to `retriever.hpp` with no other changes. The second commit creates a new `retriever.h` that warns it is deprecated and includes `retriever.hpp` as well as making the code changes to use retriever.hpp.

### CODE REVIEW

**Comment by sloretz** [comment]
CI linux build: `--packages-above resource_retriever` test: `--packages-select resource_retriever` to look for warnings downstream: [Build Status](https://ci.ros2.org/job/ci_linux/11799/)

File: resource_retriever/include/resource_retriever/retriever.h [hunk 1]
```diff
@@ -29,65 +29,15 @@
...
-#include <stdexcept>
-#include <string>
+#ifdef _MSC_VER
+#pragma message("Including header <resource_retriever/retriever.h> is deprecated,")
+#pragma message("include <resource_retriever/retriever.hpp> instead.")
+#else
+// *INDENT-OFF* (prevent uncrustify from adding indention below)
+#warning Including header <resource_retriever/retriever.h> is deprecated, \
+include <resource_retriever/retriever.hpp> instead.
+#endif
```
Comment (clalancette) [inline resource_retriever/include/resource_retriever/retriever.h:39]: Hm, do we need the separate `pragma` for Windows?  It doesn't seem like we are doing that [elsewhere](https://github.com/ros/pluginlib/blob/ros2/pluginlib/include/pluginlib/class_loader.h), but that is actually the only example I can find in our current codebase.

**Review by clalancette — COMMENTED** [review]
One minor question, and it looks like CI failed.  Once those are fixed this is fine with me, thanks for doing this.

File: resource_retriever/include/resource_retriever/retriever.h (hunk 1 above)
Comment (sloretz) [reply resource_retriever/include/resource_retriever/retriever.h:39]: I mostly blindly copied it from `class_loader` https://github.com/ros/class_loader/blob/8864be372a83921f15906ae80752be9d6a300b81/include/class_loader/class_loader_core.h#L35-L42

This Stack Overflow answer suggests the answer is yes https://stackoverflow.com/questions/171435/portability-of-warning-preprocessor-directive - maybe no code on windows has tried to include `<pluginlib/class_loader.h>`?

**Comment by sloretz** [comment]
> One minor question, and it looks like CI failed. 

Oops, I used the wrong build args. New build `--packages-above-and-dependencies resource_retriever` test `--packages-select resource_retriever` 

* Linux [Build Status](https://ci.ros2.org/job/ci_linux/11804/)
* Windows [Build Status](https://ci.ros2.org/job/ci_windows/11746/)

File: resource_retriever/include/resource_retriever/retriever.h (hunk 1 above)
Comment (clalancette) [reply resource_retriever/include/resource_retriever/retriever.h:39]: > This Stack Overflow answer suggests the answer is yes https://stackoverflow.com/questions/171435/portability-of-warning-preprocessor-directive - maybe no code on windows has tried to include `<pluginlib/class_loader.h>`?

Ah, interesting.  OK, totally fine with me, just missed where this came from.

**Review by clalancette — APPROVED** [review]

**Comment by sloretz** [comment]
CI (build: `--packages-above-and-dependencies resource_retriever` test: ` --packages-select resource_retriever rviz_common rviz_default_plugins rviz_rendering rviz_rendering_tests`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11821/)
  * java traceback, rebuilding: [Build Status](https://ci.ros2.org/job/ci_linux/11829/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6892/)
  * java traceback, rebuilding: [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/6896/)
    * java traceback, rebuilding: [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/6897/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9648/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11755/)

**Pull request merged**